File: Platform/Maml.Wasm/wwwroot/Input.ts
import { wasm } from "./wasm.js";

type InputWasm = {
	HandlePointerMove: (x: number, y: number, buttonMask: number) => void;
	HandlePointerDown: (x: number, y: number, button: number, buttonMask: number) => void;
	HandlePointerUp: (x: number, y: number, button: number, buttonMask: number) => void;
	HandleWheel: (x: number, y: number, dx: number, dy: number) => void;
	HandleKeyDown: (key: string, echo: boolean) => void;
	HandleKeyUp: (key: string) => void;
	HandleFocus: () => void;
	HandleBlur: () => void;
}

function makeOutput(id: string = ""): HTMLElement {
	var output = document.createElement("div");
	output.className = "output";
	output.id = id;
	output.innerHTML = `[${id}]`;
	document.getElementById("out").append(output);
	return output;
}

class Input {
	public async init() {
		wasm.setModuleImports("input.js", this);

		const input = await wasm.getAssemblyExports<InputWasm>("Maml.Wasm", "Maml.UserInput.Input");

		window.onscroll = (e: UIEvent) => {
			e.preventDefault();
			window.scrollTo(0, 0);
		};

		window.onpointermove = (e: PointerEvent) => {
			e.preventDefault();
			input.HandlePointerMove(e.clientX, e.clientY, e.buttons);
		}

		window.onpointerdown = (e: PointerEvent) => {
			e.preventDefault();
			input.HandlePointerDown(e.clientX, e.clientY, 1 << e.button, e.buttons);
		}

		window.onpointerup = (e: PointerEvent) => {
			e.preventDefault();
			input.HandlePointerUp(e.clientX, e.clientY, 1 << e.button, e.buttons);
		}

		window.ontouchstart = (e: TouchEvent) => {
			e.preventDefault();
		}

		window.onwheel = (e: WheelEvent) => {
			input.HandleWheel(e.clientX, e.clientY, e.deltaX, e.deltaY);
		}

		window.onkeydown = (e: KeyboardEvent) => {
			input.HandleKeyDown(e.key, e.repeat);
		}

		window.onkeyup = (e: KeyboardEvent) => {
			input.HandleKeyUp(e.key);
		}

		window.onfocus = (e: FocusEvent) => {
			input.HandleFocus();
		}

		window.onblur = (e: FocusEvent) => {
			input.HandleBlur();
		}
	}

	private readonly Outputs: {[key: string]: HTMLElement} = {
		WindowResize: makeOutput("out-window-resize"),
		PointerMove: makeOutput("out-pointer-move"),
		PointerButton: makeOutput("out-pointer-button"),
		Wheel: makeOutput("out-wheel"),
		Keyboard: makeOutput("out-keyboard"),
		Focus: makeOutput("out-focus")
	};

	private report(id: string, value: string): void {
		this.Outputs[id].innerHTML = value;
	}
}

export const input = new Input();

File: Platform/Maml.Wasm/UserInput/Input.cs
using Maml.Events;
using System.Runtime.InteropServices.JavaScript;

namespace Maml.UserInput;

public static partial class Input
{

	[JSExport]
	internal static void HandlePointerMove(int x, int y, int iButtonMask)
	{
		PointerButton buttonMask = (PointerButton)iButtonMask;
		PointerMove?.Invoke(new PointerEvent
		{
			Position = new(x, y),
			ButtonMask = buttonMask,
		});
	}

	[JSExport]
	internal static void HandlePointerDown(int x, int y, int iButton, int iButtonMask)
	{
		PointerButton button = (PointerButton)iButton;
		if (button == PointerButton.Right) { button = PointerButton.Middle; }
		else if (button == PointerButton.Middle) { button = PointerButton.Right; }

		PointerButton buttonMask = (PointerButton)iButtonMask;
		PointerDown?.Invoke(new PointerEvent
		{
			Position = new(x, y),
			Button = button,
			ButtonMask = buttonMask,
		});
	}

	[JSExport]
	internal static void HandlePointerUp(int x, int y, int iButton, int iButtonMask)
	{
		PointerButton button = (PointerButton)iButton;
		if (button == PointerButton.Right) { button = PointerButton.Middle; }
		else if (button == PointerButton.Middle) { button = PointerButton.Right; }

		PointerButton buttonMask = (PointerButton)iButtonMask;
		PointerUp?.Invoke(new PointerEvent
		{
			Position = new(x, y),
			Button = button,
			ButtonMask = buttonMask,
		});
	}

	[JSExport]
	internal static void HandleWheel(int x, int y, double dx, double dy)
	{
		Wheel?.Invoke(new WheelEvent
		{
			Position = new(x, y),
			Delta = new(dx, dy),
		});
	}

	[JSExport]
	internal static void HandleKeyDown(string key, bool echo)
	{
		VirtualKey vk = InputHelpers.JSKeyToVirtualKey(key);
		KeyDown?.Invoke(new KeyEvent
		{
			VirtualKey = vk,
			Echo = echo,
			Pressed = true,
		});
	}

	[JSExport]
	internal static void HandleKeyUp(string key, bool echo)
	{
		VirtualKey vk = InputHelpers.JSKeyToVirtualKey(key);
		KeyUp?.Invoke(new KeyEvent
		{
			VirtualKey = vk,
			Echo = echo,
			Pressed = false,
		});
	}

	[JSExport]
	internal static void HandleFocus()
	{
		Focus?.Invoke(new FocusEvent
		{
			Focused = true,
		});
	}

	[JSExport]
	internal static void HandleBlur()
	{
		Blur?.Invoke(new FocusEvent
		{
			Focused = false,
		});
	}

	[JSImport("report", "input.js")]
	internal static partial void Report(string id, string value);
}
